lib/WebSocket.js

```diff
@@ -25,277 +24,391 @@ const closeTimeout = 30 * 1000; // Allow 30 seconds to terminate the connection
+    if (this.readyState !== WebSocket.OPEN) {
+      if (dontFailWhenClosed) return;
+      throw new Error('not opened');
+    }
+
+    options = options || {};
+    if (options.mask === undefined) options.mask = !this._isServer;
+
+    this._sender.ping(data, options);
   }
 
-  this._sender.send(data, options, cb);
-};
+  /**
+   * Send a pong message.
+   *
+   * @param {*} data The message to send
+   * @param {Object} options Options object
+   * @param {Boolean} options.mask Indicates whether or not to mask `data`
+   * @param {Boolean} dontFailWhenClosed Indicates whether or not to throw an if the connection isn't open
+   * @public
+   */
+  pong (data, options, dontFailWhenClosed) {
+    if (this.readyState !== WebSocket.OPEN) {
+      if (dontFailWhenClosed) return;
+      throw new Error('not opened');
+    }
 
-/**
- * Immediately shuts down the connection
- *
- * @api public
- */
-WebSocket.prototype.terminate = function terminate () {
-  if (this.readyState === WebSocket.CLOSED) return;
+    options = options || {};
+    if (options.mask === undefined) options.mask = !this._isServer;
 
-  if (this._socket) {
-    this.readyState = WebSocket.CLOSING;
+    this._sender.pong(data, options);
+  }
 
-    // End the connection
-    try {
-      this._socket.end();
-    } catch (e) {
-      // Socket error during end() call, so just destroy it right now
-      cleanupWebsocketResources.call(this, true);
+  /**
+   * Send a data message.
+   *
+   * @param {*} data The message to send
+   * @param {Object} options Options object
+   * @param {Boolean} options.compress Specifies whether or not to compress `data`
+   * @param {Boolean} options.binary Specifies whether `data` is binary or text
+   * @param {Boolean} options.fin Specifies whether the fragment is the last one
+   * @param {Boolean} options.mask Specifies whether or not to mask `data`
+   * @param {Function} cb Callback which is executed when data is written out
+   * @public
+   */
+  send (data, options, cb) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+
+    if (this.readyState !== WebSocket.OPEN) {
+      if (cb) cb(new Error('not opened'));
+      else throw new Error('not opened');
       return;
     }
 
-    // Add a timeout to ensure that the connection is completely
-    // cleaned up within 30 seconds, even if the clean close procedure
-    // fails for whatever reason
-    // First cleanup any pre-existing timeout from an earlier "terminate" call,
-    // if one exists.  Otherwise terminate calls in quick succession will leak timeouts
-    // and hold the program open for `closeTimout` time.
-    if (this._closeTimer) { clearTimeout(this._closeTimer); }
-    this._closeTimer = setTimeout(cleanupWebsocketResources.bind(this, true), closeTimeout);
-  } else if (this.readyState === WebSocket.CONNECTING) {
-    cleanupWebsocketResources.call(this, true);
-  }
-};
+    if (!data) data = '';
```

Comment:
@3rd-Eden I think that dropping Node 4 support is a bit too much. For example AWS Lambda does not support Node 6 yet.

Now using ws with Lambda does not make much sense 😄  but I think you get my point.